{{cookiecutter.project_slug}}/{{cookiecutter.project_slug}}/users/tests/async_server.py

```diff
@@ -0,0 +1,40 @@
+import asyncio
+import functools
+import threading
+import time
+from contextlib import contextmanager
+
+from uvicorn.config import Config
+from uvicorn.main import ServerState
+from uvicorn.protocols.http.h11_impl import H11Protocol
+from uvicorn.protocols.websockets.websockets_impl import WebSocketProtocol
+
+
+def run_loop(loop):
+    loop.run_forever()
+    loop.close()
+
+
+@contextmanager
+def run_server(app, path="/"):
```

Comment:
This seems like a lot of boilerplate 😞 Could we extract the main logic into smaller async functions and use [`pytest-asyncio`](https://pypi.org/project/pytest-asyncio/) to test them?

Reply:
yeah that was my initial thought too while developing this. But in my own project code, I do use pytest-asyncio but for a different purpose.

When a `text` is transmitted to the server, I categorize it based on the JSON passed through. You can see this every time GitHub sends a notification to you, the client. It's JSON encoded. For example, GitHub may send:

```
{"subscribe": "large base 64"}
```

What we'd do is do a switch/case with that key/value "subscribe" (e.g. this is how GitHub auto updates this GitHub issue whenever something new happens). The problem then becomes how many different categories do you want to manage. That's when I split it into a coroutine; when you split it into a coroutine, that's when you can properly "unit" test.